ecosystem/purescript-0.14-migration-guide.md

```diff
@@ -0,0 +1,622 @@
+
+```purs
+data Proxy a
+```
+
+This previously had the `Type`-defaulted kind `Type -> Type`. Now this will be generalized to `forall k. k -> Type`. It is better to be explicit about polymorphism by writing signatures, and the compiler will warn about missing kind signatures when it infers polymorphic kinds.
+
+To avoid the compiler warning, provide a kind signature:
+
+```purs
+data Proxy :: forall k. k -> Type
+data Proxy a = Proxy
+```
+
+Classes can have signatures too, and they must end with a `Constraint` kind instead of `Type`. For example, here's the new definition of `Prim.Row.Cons`:
+
+```purs
+class Cons :: forall k. Symbol -> k -> Row k -> Row k -> Constraint
+class Cons label a tail row | label a tail -> row, label row -> a tail
+```
+
+### Coercible & Role Annotations
+
+PureScript 0.14 introduces a new compiler-solved class, `Prim.Coerce.Coercible`, to relate types with the same runtime representation. You can now use `Safe.Coerce.coerce` from the new `safe-coerce` library in place of `Unsafe.Coerce.unsafeCoerce` to turn an `a` into a `b` when `Coercible a b` holds.
+
+If you are currently using `unsafeCoerce` to transform two types with the same runtime representation, you should consider replacing it with `coerce`.
+
+The `Coercible` implementation introduces role annotations for type parameters (to learn more about roles, see the compiler release notes). Role annotations can loosen or strengthen the compiler-inferred roles of type parameters.
+
+For example, the compiler infers nominal roles for foreign data types, which is safe but can be too constraining. For example, we should be able to coerce `Effect Age` to `Effect Int` if `Age` is a newtype over `Int`, but because `Effect` is a foreign data type its type parameter is inferred with a nominal role. We can loosen the role with an annotation:
+
+```purs
+foreign import data Effect :: Type -> Type
+
+-- this is a role annotation, setting Effect's type parameter to be
+-- representational instead of nominal (its inferred role).
+type role Effect representational
+```
+
+In contrast, it is unsafe to allow coercions between `Map k1 a` and `Map k2 a`, even when `Coercible k1 k2` holds. We can annotate `Map`'s key type parameter with a nominal role to prevent coercions:
+
+```purs
+data Map k v = ...
+
+-- this is a role annotation, setting `Map`'s key type parameter to
+-- be nominal instead of representational (its inferred role).
+type role Map nominal representational
+```
+
+### Other Changes
+
+Primes are no longer allowed in JavaScript FFI definitions because PureScript 0.15 will generate ES modules.
+
+**To fix:**
+
+Rename any functions that contain a prime and are implemented via the FFI:
+
+```js
+// before
+exports["functionName'"] = function (a) { return a; }
+
+// after: these are common options
+exports.functionNameImpl = function (a) { return a; }
+exports._functionName = function (a) { return a; }
+exports.functionNamePrime = function (a) { return a; }
```

Comment:
I find the "Prime" suffix a bit odd, personally. Have we seen this in the wild?

Reply:
No, I haven’t. I can remove it.

Reply:
Actually come to think of it, the leading underscore is probably not ideal either since it'll silence the unused variable warnings (which I would quite like to try to land soon after 0.14.0). https://github.com/purescript/purescript/pull/3819 Perhaps we should use a trailing underscore instead? Or maybe only suggest `Impl`?

Reply:
Either one seems fine to me, but we should explicitly state the leading underscore is not recommended.